src/Paprika/Store/PagedDb.MultiHeadChain.cs

```diff
@@ -0,0 +1,977 @@
+﻿using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Threading.Channels;
+using Paprika.Chain;
+using Paprika.Crypto;
+using Paprika.Data;
+using Paprika.Utils;
+
+namespace Paprika.Store;
+
+public sealed partial class PagedDb
+{
+    public IMultiHeadChain OpenMultiHeadChain()
+    {
+        lock (_batchLock)
+        {
+            if (_batchCurrent != null)
+            {
+                ThrowOnlyOneBatch();
+            }
+
+            var chain = new MultiHeadChain(this);
+            _batchCurrent = chain;
+
+            return chain;
+        }
+    }
+
+    private class MultiHeadChain : IMultiHeadChain
+    {
+        private readonly PagedDb _db;
+        private readonly BufferPool _pool;
+
+        // Batches grouped by id and number
+        private readonly Dictionary<uint, List<ProposedBatch>> _proposedBatchesByBatchId = new();
+        private readonly Dictionary<Keccak, ProposedBatch> _proposedBatchesByHash = new();
+
+        // Readers
+        private readonly Dictionary<Keccak, Reader> _readers = new();
+        private readonly Queue<Reader> _readersDisposalQueue = new();
+        private Reader _lastFinalized;
+        private readonly ReaderWriterLockSlim _readerLock = new();
+
+        // Proposed batches that are finalized
+        private readonly HashSet<Keccak> _beingFinalized = new();
```

Comment:
With this new approach, we are storing a lot of data in-memory and this is not fetched from buffer pool/existing set of DB pages. Couple of general questions:
1.How big can this grow? What does benchmarks show?
2. What possible worst-case scenario can arise in case of a crash, since we would lose everything in-memory which had not been checkpointed?

Reply:
1. We copy/store only pages that are written. This means roughly ~2000 pages per block which gives 8MB per block. This, multiplied the estimated number of not finalized blocks set to ~64 gives us ~500MB of memory. From what I recall when running a node with it, it was much lower though.
2. We lose what was not finalized. The worst scenario is that once the node is restarted it gets the last remembered hash and processes what is missing. This is not different from the current one.